### PR TITLE
Registration: Add server script, style block type property

### LINKED ISSUE #2756: Blocks: Revisit pattern for enqueuing scripts and styles
Related: #2751

With the proposal in #2751 to require a PHP block definition for all blocks, it becomes reasonable to revisit whether scripts and styles should be enqueued separately from the block definition itself. Notably, this could have an impact on the front-end display of a block, as currently we must enqueue all scripts and styles for all blocks, regardless of whether a block of that particular type exists on the current screen. If the script and style handles were defined with the block, it is possible to limit this to only the blocks present in the current screen. This would, however, require that the post contents be parsed prior to display to detect blocks present in the post, to then check against all registered block types and associated script and styles handles. We should measure the performance impact of this parsing in relation to the benefit associated with filtering scripts.

_Before:_

```php
function myplugin_enqueue_block_editor_assets() {
	wp_enqueue_script(
		'nice-block',
		plugins_url( 'block.js', __FILE__ ),
		array( 'wp-blocks', 'wp-components', 'wp-element' )
	);
}
add_action( 'enqueue_block_editor_assets', 'myplugin_enqueue_block_editor_assets' );
```

_After:_

```php
function myplugin_enqueue_block_editor_assets() {
	wp_register_script(
		'nice-block',
		plugins_url( 'block.js', __FILE__ ),
		array( 'wp-blocks', 'wp-components', 'wp-element' )
	);

	register_block( 'nice-block', array(
		'script_handle' => 'nice-block',

		// ...
	) );
}
add_action( 'init', 'myplugin_enqueue_block_editor_assets' );
```

Alternatively, these files could be "discovered" adjacent the block registering file by naming conventions (e.g. find `style.css`, `edit.css`, `block.js` adjacent `myblock/index.php`). This could encourage developers to define their blocks in a standalone folder/file structure for easier separation overview.

### PR DESCRIPTION
**Pull request by aduth**
Closes #2756
Related: #2751

This pull request seeks to implement and document an alternative approach to enqueueing block assets, integrating new settings into the `register_block_type` server-side API to help manage context-specific script and style enqueues.

The `register_block_type` API now supports the following properties:

- `editor_script`
- `editor_style`
- `script`
- `style`

Assigning one or more of these to registered script or style handles will cause them to be enqueued in the corresponding render contexts (editor-specific or always).

This is intended as part of an effort to coalesce initial block implementations toward the server-side block registration API, as proposed in #2751 to be the default for new blocks.

__Testing instructions:__

Since this simply builds on existing APIs and no core blocks have been ported to using it (unclear whether this is a direction we'd want to go), there should be no expected impact of these changes.

You may try including a third-party block code and changing to implement with the `register_block_type` API.

Example:

https://gist.github.com/pento/19b35d621709042fc899e394a9387a54

```diff
diff --git a/stars-block.php b/stars-block.php
index da5598c..9ea8cfd 100755
--- a/stars-block.php
+++ b/stars-block.php
@@ -9,15 +9,6 @@
  * Description: Everyone loves stars! Let's add stars (now with more meta)!
  */
 
-function stars_block_enqueue_block_editor_assets() {
-       wp_enqueue_script(
-               'stars-block',
-               plugins_url( 'stars-block.js', __FILE__ ),
-               array( 'wp-blocks' )
-       );
-}
-add_action( 'enqueue_block_editor_assets', 'stars_block_enqueue_block_editor_assets' );
-
 function stars_block_init() {
        // Register the postmeta key that we'll store our data in
        register_meta( 'post', 'stars', array(
@@ -25,5 +16,15 @@ function stars_block_init() {
                'single'       => true,
                'type'         => 'number',
        ) );
+
+       wp_register_script(
+               'stars-block',
+               plugins_url( 'stars-block.js', __FILE__ ),
+               array( 'wp-blocks' )
+       );
+
+       register_block_type( 'stars/stars-block', array(
+               'editor_script' => 'stars-block',
+       ) );
 }
 add_action( 'init', 'stars_block_init' );
```

### CODE REVIEW

File: docs/blocks-stylesheet.md [hunk 1]
```diff
@@ -61,47 +61,54 @@ The class name is generated using the block's name prefixed with `wp-block-`, re
...
 
 ```php
 <?php
 
-function gutenberg_boilerplate_enqueue_block_editor_assets() {
-	wp_enqueue_script(
-		'gutenberg-boilerplate-es5-step02',
+function gutenberg_boilerplate_block() {
+	wp_register_script(
+		'gutenberg-boilerplate-es5-step02-editor-script',
```
Comment (aduth) [inline docs/blocks-stylesheet.md]: A seemingly basic question I wasn't sure of: Are handles required to be unique only between script and style registrations, or for both? i.e. Is it valid to:

```php
wp_register_script( 'foo', /* ... */ );
wp_register_style( 'foo', /* ... */ );
```

**Comment by gziolo** [comment]
We will have to update WP-CLI to promote those changes. Please ping me when it is ready to merge so I could open a follow-up PR.

File: docs/blocks-stylesheet.md (hunk 1 above)
Comment (westonruter) [reply docs/blocks-stylesheet.md]: They are free to be the same. I do it all the time, and core does it as well. For example, `customize-controls`:

* Script: https://github.com/WordPress/wordpress-develop/blob/53f1c74e1c55910f9b7e4aebc8033571d3e52df8/src/wp-includes/script-loader.php#L586
* Style: https://github.com/WordPress/wordpress-develop/blob/53f1c74e1c55910f9b7e4aebc8033571d3e52df8/src/wp-includes/script-loader.php#L1065

Comment (aduth) [reply docs/blocks-stylesheet.md]: >They are free to be the same.

Thanks! Updated in 5972c9b21ef3314c2cf91fc620de930bf97d860f.

**Comment by aduth** [comment]
@gziolo I expect this is ready to merge as-is. Do you have any work you'd like to have prepared before I do so?

**Comment by gziolo** [comment]
It looks like non-breaking change so you can merge anytime. I will update WP-CLI next week to use this simplified API 👍

**Comment by gziolo** [comment]
@aduth I created https://github.com/wp-cli/scaffold-command/pull/111 for `wp scaffold block` to include your changes. The output looks great 👍

**Comment by aduth** [comment]
Investigating breakage for meta block attributes, finding via bisect that somehow this pull request (specifically 2c80e7bd1c3cd83e9e39166428c40da4d8b2b909) broke it...

**Pull request merged**